### PR TITLE
Const correctness for mesh partition

### PR DESCRIPTION
**Pull request by sebastiangrimberg**
- Fixes which allow the input `partitioning` pointer to be `const` in `ParMesh`, `ParNURBSExtension`, and `ParNCMesh`
- Removes unused `partitioning_cache` member from `ParMesh` class (seems left over from https://github.com/mfem/mfem/pull/2920)

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#4190](https://github.com/mfem/mfem/pull/4190) | @sebastiangrimberg | @tzanio | @termi-official + @jandrej | 5/11/24 | 6/24/24 | 6/25/24 | |

<details>
<summary>PR Checklist</summary>
    
- [ ] Code builds.
- [ ] Code passes `make style`.
- [ ] Update `CHANGELOG`:
    - [ ] Is this a new feature users need to be aware of? New or updated example or miniapp?
    - [ ] Does it make sense to create a new section in the `CHANGELOG` to group with other related features?
- [ ] Update `INSTALL`:
    - [ ] Had a new optional library been added? If so, what range of versions of this library are required? (*Make sure the external library is compatible with our BSD license, e.g. it is not licensed under GPL!*)
    - [ ] Have the version ranges for any required or optional libraries changed?
    - [ ] Does `make` or `cmake` have a new target?
    - [ ] Did the requirements or the installation process change? *(rare)*
- [ ] Update continuous integration server configurations if necessary (e.g. with new version requirements for each of MFEM's dependencies)
    - [ ] `.github`
    - [ ] `.appveyor.yml`
- [ ] Update `.gitignore`:
    - [ ] Check if `make distclean; git status` shows any files that were generated from the source by the project (not an IDE) but we don't want to track in the repository.
    - [ ] Add new patterns (just for the new files above) and re-run the above test.
- [ ] New examples:
    - [ ] All sample runs at the top of the example source file work.
    - [ ] Update `examples/makefile`:
      - [ ] Add the example code to the appropriate `SEQ_EXAMPLES` and `PAR_EXAMPLES` variables.
      - [ ] Add any files generated by it to the `clean` target.
      - [ ] Add the example binary and any files generated by it to the top-level `.gitignore` file.
    - [ ] Update `examples/CMakeLists.txt`:
      - [ ] Add the example code to the `ALL_EXE_SRCS` variable.
      - [ ] Make sure `THIS_TEST_OPTIONS` is set correctly for the new example.
   - [ ] List the new example in `doc/CodeDocumentation.dox`.
   - [ ] If new examples directory (e.g.`examples/pumi`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
      - [ ] Update or add example-specific documentation, see e.g. the `src/examples.md`.
      - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
      - [ ] In `examples.md`, list the example under the appropriate categories, add new categories if necessary.
      - [ ] Add a short description of the example in the "Extensive Examples" section of `features.md`.
- [ ] New miniapps:
   - [ ] All sample runs at the top of the miniapp source file work.
   - [ ] Update top-level `makefile` and `makefile` in corresponding miniapp directory.
   - [ ] Add the miniapp binary and any files generated by it to the top-level `.gitignore` file.
   - [ ] Update CMake build system:
      - [ ] Update the `CMakeLists.txt` file in the `miniapps` directory, if the new miniapp is in a new directory.
      - [ ] Add/update the `CMakeLists.txt` file in the new miniapp directory.
      - [ ] Consider adding a new test for the new miniapp.
   - [ ] List the new miniapp in `doc/CodeDocumentation.dox`
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), add it to `MINIAPP_SUBDIRS` in the `makefile`.
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
     - [ ] Update or add miniapp-specific documentation, see e.g. the `src/meshing.md` and `src/electromagnetics.md` files.
     - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
     - [ ] The miniapps go at the end of the page, and are usually listed only under a specific "Application (PDE)" category.
     - [ ] Add a short description of the miniapp in the "Extensive Examples" section of `features.md`.
- [ ] New capability:
   - [ ] All new public, protected, and private classes, methods, data members, and functions have full Doxygen-style documentation in source comments. Documentation should include descriptions of member data, function arguments and return values, template parameters, and prerequisites for calling new functions.
   - [ ] Pointer arguments and return values must specify whether ownership is being transferred or lent with the call.
   - [ ] Any new functions should include descriptions of their intended use e.g. for internal use only, user-facing, etc., along with references to example code whenever possible/appropriate.
   - [ ] Consider adding new sample runs in existing examples to highlight the new capability.
   - [ ] Consider saving cool simulation pictures with the new capability in the Confluence gallery (LLNL only) or submitting them, via pull request, to the gallery section of the `mfem/web` repo.
   - [ ] If this is a major new feature, consider mentioning it in the short summary inside `README` *(rare)*.
   - [ ] List major new classes in `doc/CodeDocumentation.dox` *(rare)*.
- [ ] Update this checklist, if the new pull request affects it.
- [ ] Run `make unittest` to make sure all unit tests pass.
- [ ] Run the tests in `tests/scripts`.
- [ ] (LLNL only) After merging:
   - [ ] Update internal tests to include the new features.
</details>

### CODE REVIEW

**Comment by tzanio** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

File: mesh/pmesh.cpp [hunk 1]
```diff
@@ -103,13 +103,13 @@ ParMesh& ParMesh::operator=(ParMesh &&mesh)
...
 
-ParMesh::ParMesh(MPI_Comm comm, Mesh &mesh, int *partitioning_,
+ParMesh::ParMesh(MPI_Comm comm, Mesh &mesh, const int *partitioning_,
                  int part_method)
    : glob_elem_offset(-1)
    , glob_offset_sequence(-1)
    , gtopo(comm)
 {
-   int *partitioning = NULL;
+   const int *partitioning = partitioning_ ? partitioning_ : nullptr;
```
Comment (termi-official) [inline mesh/pmesh.cpp:112]: I think we can eliminate this line and rename `partitioning_` to `partitioning`.

File: mesh/pmesh.hpp [hunk 2]
```diff
@@ -343,7 +343,7 @@ class ParMesh : public Mesh
        meshes and quick space-filling curve equipartitioning for nonconforming
        meshes (elements of nonconforming meshes should ideally be ordered as a
        sequence of face-neighbors). */
-   ParMesh(MPI_Comm comm, Mesh &mesh, int *partitioning_ = NULL,
+   ParMesh(MPI_Comm comm, Mesh &mesh, const int *partitioning_ = NULL,
```
Comment (termi-official) [inline mesh/pmesh.hpp:346]: For consistency with the remaining variables, can we rename partitioning_ to partitioning and use nullptr?

File: mesh/pncmesh.hpp [hunk 3]
```diff
@@ -67,7 +67,7 @@ class ParNCMesh : public NCMesh
    /// Construct by partitioning a serial NCMesh.
    /** SFC partitioning is used by default. A user-specified partition can be
        passed in 'part', where part[i] is the desired MPI rank for element i. */
-   ParNCMesh(MPI_Comm comm, const NCMesh& ncmesh, int* part = NULL);
+   ParNCMesh(MPI_Comm comm, const NCMesh& ncmesh, const int* part = NULL);
```
Comment (termi-official) [inline mesh/pncmesh.hpp:70]: For consistency with the remaining variables, can we rename part to partitioning and use nullptr?

**Review by termi-official — APPROVED** [review]
Some stylistic suggestions. LGTM otherwise.

File: mesh/pmesh.hpp (hunk 2 above)
Comment (termi-official) [reply mesh/pmesh.hpp]: But partitioning is not shadowed here, because it is not a member of the class, right?

Comment (sebastiangrimberg) [reply mesh/pmesh.hpp]: `partitioning` is still used in the local scope so I think keeping this as `partitioning_` is needed. See  [aab273b](https://github.com/mfem/mfem/pull/4190/commits/aab273b303873553f6d7533f4491cbcc3c54d608) which addressed [this comment](https://github.com/mfem/mfem/pull/4190#discussion_r1598081303) but also made it so the local `partitioning` variable is needed.

File: mesh/pmesh.cpp (hunk 1 above)
Comment (sebastiangrimberg) [reply mesh/pmesh.cpp]: Discussion continued in https://github.com/mfem/mfem/pull/4190#discussion_r1607117726.

File: mesh/pncmesh.hpp (hunk 3 above)
Comment (sebastiangrimberg) [reply mesh/pncmesh.hpp]: Resolved in https://github.com/mfem/mfem/pull/4190/commits/3e5e18797ca8427697547b8f9d1bc46815176ce9#diff-f34a87a4063ff4ad77110d8e062b2fe3fabd7598d40dbc5f39d8b9c81d1eb0ff.

File: mesh/pmesh.hpp (hunk 2 above)
Comment (termi-official) [reply mesh/pmesh.hpp]: Oh, I see. Then it's okay. Thanks for pointing this out.

**Review by jandrej — APPROVED** [review]

**Comment by tzanio** [comment]
Merged in `next` for testing...

**Pull request merged**